File: 127-word-ladder/word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList: return 0
        
        mapper = defaultdict(list)
        wordList.append(beginWord)
        for word in wordList:
            for i in range(len(word)):
                temp = mapper.get(word[:i] + "*" + word[i+1:],[])
                temp.append(word)
                mapper[word[:i] + "*" + word[i+1:]]=temp
        
        q = deque([beginWord])
        visited = set([beginWord])
        r = 1
        while q:
            for _ in range(len(q)):
                temp = q.popleft()
                if temp == endWord: return r
                for i in range(len(temp)):
                    if temp[:i] + "*" + temp[i+1:] in mapper:
                        for new in mapper[temp[:i] + "*" + temp[i+1:]]:
                            if new not in visited:
                                q.append(new)
                                visited.add(new)
            r += 1
        return 0

File: 127-word-ladder/word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if (!wordList.contains(endWord)) return 0;

        HashMap<String, List<String>> mapper = new HashMap<String, List<String>>();
        wordList.add(beginWord);

        for (String word: wordList){
            for (int i = 0; i < word.length(); i++){
                List<String> temp = mapper.getOrDefault(word.substring(0, i) + '*' + word.substring(i + 1, word.length()), new ArrayList<String>());
                temp.add(word);
                mapper.put(word.substring(0, i) + '*' + word.substring(i + 1, word.length()), temp);
            }
        }

        Queue<String> q = new LinkedList<String>();
        HashSet<String> visited = new HashSet<String>();
        q.add(beginWord);
        visited.add(beginWord);
        int result = 1;

        while(q.size()>0){
            int l = q.size();
            for(int i = 0; i < l; i++){
                String temp = q.poll();

                if (temp.equals(endWord)) return result;
                for (int j = 0; j < temp.length(); j++){
                    for (String t: mapper.getOrDefault(temp.substring(0, j) + "*" + temp.substring(j + 1, temp.length()), new ArrayList<String>())){
                        if (!visited.contains(t)){
                                q.add(t);
                                visited.add(t);
                        }
                    }
                }
            }

            result += 1;
        }
        return 0;
    }
}